git/client.go

```diff
@@ -445,6 +448,46 @@ func (c *Client) SetBranchConfig(ctx context.Context, branch, name, value string
 	return err
 }
 
+// PushDefault returns the value of push.default in the config. If the value
+// is not set, it returns "simple" (the default value).
+func (c *Client) PushDefault(ctx context.Context) (string, error) {
+	pushDefault, err := c.Config(ctx, "push.default")
+	if err == nil {
+		return pushDefault, nil
+	}
+
+	var gitError *GitError
+	if ok := errors.As(err, &gitError); ok && gitError.ExitCode == 1 {
+		return "simple", nil
+	}
+	return "", err
+}
+
+// RemotePushDefault returns the value of remote.pushDefault in the config. If
+// the value is not set, it returns an empty string.
+func (c *Client) RemotePushDefault(ctx context.Context) (string, error) {
+	remotePushDefault, err := c.Config(ctx, "remote.pushDefault")
+	if err == nil {
+		return remotePushDefault, nil
+	}
+
+	var gitError *GitError
+	if ok := errors.As(err, &gitError); ok && gitError.ExitCode == 1 {
+		return "", nil
+	}
+
+	return "", err
+}
+
+// ParsePushRevision gets the value of the @{push} revision syntax
+// An error here doesn't necessarily mean something are broke, but may mean that the @{push}
+// revision syntax couldn't be resolved, such as in non-centralized workflows with
+// push.default = simple. Downstream consumers should consider how to handle this error.
```

Comment:
Aside from a few minor grammar corrections, I'm a little uneasy with an error being raised here that requires the user to discern whether it is actually a problem or not.

Any other guidance we should capture here to lead others to scenarios where this isn't an error?

```suggestion
// ParsePushRevision gets the value of the @{push} revision syntax.
// An error here doesn't necessarily mean something is broken, but may mean that the @{push}
// revision syntax couldn't be resolved, such as in non-centralized workflows with
// push.default = simple. Downstream consumers should consider how to handle this error.
```

Reply:
Man, writing -> editing -> rewriting again and again really makes grammar hard... Thanks for catching this!

As for the guidance, being helpful may be difficult, though not impossible. The biggest issue that may happen with this is that the `git` call fails for some unknown reason.

However, there are situations where `@{push}` won't resolve but we can still figure out the PRRefs. One of these cases is having `push.default = simple` while configured to a triangular workflow. In this case, we get a specific `@{push}` error:

```
// git config
[push]
        default = simple
[branch "triangular-branch"]
	remote = upstream
	merge = refs/heads/main

$ git rev-parse --abbrev-ref triangular-branch@{push}
fatal: cannot resolve 'simple' push to a single destination
```

This same command gives another error for a common triangular workflow setup:

```
// git config:
[push]
        default = current
[branch "triangular-branch"]
	remote = upstream
	merge = refs/heads/main
	pushremote = origin

// Command
$ git rev-parse --abbrev-ref triangular-branch@{push}
triangular-branch@{push}
fatal: ambiguous argument 'triangular-branch@{push}': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

Git exits `128` in both of these cases, but we can still figure out the PRRefs from the provided configs. 

Now, there's a decent chance I'm missing something about these triangular workflow setups, but I'm not sure what other errors this may produce that we can't recover from. I suppose we could surface them and learn about them that way, but letting the code that needs the push revision decide what to do when it doesn't get it (and instead gets an error) seems right to me...

What do you think, given that context?

Reply:
Thank you for explaining in detail!  The examples really helped.

I suppose lets just move forward and we deal with any problems that arise. 🤷 Eventually, I hope we have a bit more documentation on various development patterns and concerns for ourselves and contributors to consult, which may be the better place to highlight this particular scenario. 👍